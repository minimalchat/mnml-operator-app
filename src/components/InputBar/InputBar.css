.InputBar {
  display: flex;
  background: #fff;
  min-height: 50px;
}

.InputBar input {
  flex-grow: 1;
  outline: 0;
  display: flex;
  padding: 1rem;
  resize: none;
  border: none;
}

.InputBar .Button--send {
  margin: 4px;
}
