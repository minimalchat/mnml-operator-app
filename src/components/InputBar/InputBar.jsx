import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Button from '../Button/Button.jsx';
import { sendMessage, operatorTyping } from '../../store/Chat';
import './InputBar.css';

// TODO: Move to a constants file
const KEY_ENTER = 13;

class InputBar extends Component {
  static propTypes = {
    operator: PropTypes.string,
    activeId: PropTypes.string,
    dispatch: PropTypes.func.isRequired,
  }

  static defaultProps = {
    operator: '',
    activeId: '',
  }

  state = {
    chatText: '',
  }

  onKeyPress = (e) => {
    const {
      key, keyCode, shiftKey, ctrlKey, altKey,
    } = e;
    const {
      dispatch, activeId, operator,
    } = this.props;
    console.log(`INPUT KEYPRESS ${key} (${keyCode}), SHIFT ${shiftKey}, CTRL ${ctrlKey}, ALT ${altKey}`);

    if (keyCode === KEY_ENTER) {
      if (!shiftKey) {
        console.log('SENDING MESSAGE ...');

        this.sendChat();
        this.setState({ [e.target.name]: '' });

        e.preventDefault();
      }
    } else {
      dispatch(operatorTyping(this.formatMessage(null, operator, activeId)));
    }
  }

  formatMessage = (content, operatorID, chatID) => ({
    timestamp: (new Date()).toISOString(),
    author: `operator.${operatorID}`,
    content,
    chat: chatID,
  })

  // Dummy function to stub out sending message via socket
  //  this will eventually happen via redux actions.
  sendChat = () => {
    const { chatText } = this.state;
    const { dispatch, activeId, operator } = this.props;

    dispatch(sendMessage(this.formatMessage(chatText, operator, activeId)));
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  render () {
    const { chatText } = this.state;
    return (
      <div className="InputBar">
        <input
          onChange={this.handleChange}
          onKeyDown={this.onKeyPress}
          name="chatText"
          value={chatText}
          placeholder="Say something meaningful &hellip;"
        />
        <Button variant="send" onClick={this.sendChat}>Send</Button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  activeId: state.chat.activeId,
  operator: state.config.operator,
});

const mapDispatchToProps = dispatch => ({
  dispatch,
});


export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(InputBar);
