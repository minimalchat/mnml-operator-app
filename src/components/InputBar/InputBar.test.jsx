import React from 'react';

import { shallow } from 'enzyme';

import InputBar from './InputBar.jsx';

const store = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
  getState: jest.fn(() => ({
    chat: {
      activeId: 'TEST',
      config: {
        operator: 'TEST',
      },
    },
  })),
};

describe('InputBar', () => {
  it('matches snapshot', () => {
    const component = shallow(<InputBar store={store} socket={{}} />);

    expect(component).toMatchSnapshot();
  });
});
