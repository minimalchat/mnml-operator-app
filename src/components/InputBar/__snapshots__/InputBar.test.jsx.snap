// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBar matches snapshot 1`] = `
<InputBar
  activeId="TEST"
  dispatch={[MockFunction]}
  operator="TEST"
  socket={Object {}}
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Object {
              "chat": Object {
                "activeId": "TEST",
              },
              "config": Object {
                "operator": "TEST",
              },
            },
          },
          Object {
            "isThrow": false,
            "value": Object {
              "chat": Object {
                "activeId": "TEST",
              },
              "config": Object {
                "operator": "TEST",
              },
            },
          },
        ],
      },
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
/>
`;
